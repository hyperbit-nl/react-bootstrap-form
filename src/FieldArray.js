import React from 'react';
import { FieldArray } from 'formik';
import { Button } from 'react-bootstrap';
import { getLayout, getSectionLabel, RBArrayAssignProps, RBCustomAssignProps, RBFieldAssignProps, RBFormObjectRecursiveAssignProps } from './FieldCommon';
import { RBField } from './Field';

export const RBFieldArray = (props) => {
  const {name, ...rest} = props; // Make sure to remove 'name' form the rest.
  const {values, errors, didSubmbit, handleChange, initialValues, addButton, noRowsMessage, children, as, label} = rest;

  const Component = (
    <FieldArray {...{name}} render={({push, remove}) => {
      const addFunction = () => (push(initialValues));

      const Add = (addButton === undefined || addButton == true) ? (
        <Button className="mb-3" variant="primary" type="button" onClick={addFunction}>Add</Button>
      ) : (
        !addButton ? null : addButton(addFunction)
      );

      return (
        <>
          {values && values.length > 0 ? (
            values.map((_, indexValue) => (
              <React.Fragment key={indexValue}>
                {RBFormObjectRecursiveAssignProps(children, (child) => {
                  const {type, props} = child;
                  const {type: typeValue, nameValue, customInput} = props;
                  const typeComponent = customInput ? 'custom' : type;

                  switch (typeComponent) {
                    case RBField:
                      return RBFieldAssignProps({
                        name: nameValue,
                        type: typeValue,
                        values,
                        errors,
                        didSubmbit,
                        handleChange,
                        nameArray: name,
                        indexValue,
                      }, true);

                    case RBFieldArray:
                      return RBArrayAssignProps({
                        name: nameValue,
                        values,
                        errors,
                        didSubmbit,
                        handleChange,
                        initialValues,
                        nameArray: name,
                        indexValue,
                      }, true);

                    case 'custom':
                      return RBCustomAssignProps({
                        name: nameValue,
                        type: typeValue,
                        values,
                        errors,
                        didSubmbit,
                        handleChange,
                        nameArray: name,
                        indexValue,
                      }, true);

                    default: {
                      if (props['array-remove']) {
                        return {
                          ['array-remove']: 'true',
                          onClick: () => {
                            remove(indexValue);
                          },
                        };
                      }
                      return;
                    }
                  }
                })}
              </React.Fragment>
            ))
          ) : (
            noRowsMessage
          )}
          {Add}
        </>
      );
    }} />
  );

  return getLayout({as}, label && getSectionLabel(rest), Component, rest);
};
