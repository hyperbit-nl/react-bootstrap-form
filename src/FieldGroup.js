import { getIdLabel, getLayout, getSectionLabel } from './FieldCommon';

export const RBFieldGroup = (props) => {
  const {id, label, hiddenLabel, as, children} = props;

  const propsGroup = {id, role: 'group', as};
  const propsLabel = {...props, name: id, className: 'rbf-group'};

  let Label;

  if (!hiddenLabel) {
    Label = getSectionLabel(propsLabel);
    propsGroup['aria-labelledby'] = getIdLabel(id);

  } else {
    propsGroup['aria-label'] = label;
  }

  return getLayout(propsGroup, Label, children, props);
};
