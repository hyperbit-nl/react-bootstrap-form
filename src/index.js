export {
  RBForm as default
} from './Form';

export {
  RBFieldGroup as FieldGroup
} from './FieldGroup';

export {
  RBFieldArray as FieldArray
} from './FieldArray';

export {
  RBField as Field
} from './Field';
