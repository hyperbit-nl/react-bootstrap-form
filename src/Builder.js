import React, { useEffect } from 'react';
import { Button, Form, Spinner } from 'react-bootstrap';
import { RBArrayAssignProps, RBCustomAssignProps, RBFieldAssignProps, RBFormObjectRecursiveAssignProps } from './FieldCommon';
import { RBField } from './Field';
import { RBFieldArray } from './FieldArray';

const createSubmitSpinner = () => (
  <Spinner className="ml-2 align-middle" animation="border" variant="primary" role="status">
    <span className="sr-only">Waiting for response</span>
  </Spinner>
);

const createSubmit = (sendText, {isLoadingSubmit, errorSubmit, successSubmit, didSubmbit, isValid, validationErrorText}) => (
  <div>
    <Button disabled={isLoadingSubmit} variant="primary" type="submit">{sendText}</Button>
    {isLoadingSubmit && createSubmitSpinner()}
    {!errorSubmit && !successSubmit && didSubmbit && !isValid ? (
      <span className="ml-2 rbf-text-red" role="status">{validationErrorText}</span>
    ) : !isLoadingSubmit ? (
      errorSubmit && (
        <span className="ml-2 rbf-text-red" role="status">{errorSubmit}</span>
      )
    ) ?? (
      successSubmit && (
        <span className="ml-2 rbf-text-green" role="status">{successSubmit}</span>
      )
    ) : null}
  </div>
);

const getSubmit = (customSubmit, props) => {
  if (customSubmit) {
    if (typeof customSubmit === 'string') {
      return createSubmit(customSubmit, props);
    }

    const {isLoadingSubmit, successSubmit, errorSubmit} = props;
    return customSubmit(isLoadingSubmit, successSubmit, errorSubmit);
  }

  return createSubmit('Send', props);
};

export const RBBuilder = ({
  form: {
    validateForm,
    handleSubmit,
    values,
    setFieldValue: handleChange,
    errors,
    setErrors,
    submitCount,
    isValid
  },
  validationSchema,
  initialValues,
  customSubmit,
  isLoadingSubmit,
  errorSubmit,
  fieldErrorSubmit,
  successSubmit,
  validationErrorText = 'Error',
  children
}) => {
  const didSubmbit = submitCount > 0;

  useEffect(() => {
    if (didSubmbit) {
      validateForm();
    }
  }, [didSubmbit, validateForm, validationSchema]);

  useEffect(() => {
    if (fieldErrorSubmit) {
      setErrors(fieldErrorSubmit);
    }
  });

  return (
    <Form
      noValidate
      onSubmit={handleSubmit}
    >
      {RBFormObjectRecursiveAssignProps(children, ({type, props}) => {
        const typeComponent = props.customInput ? 'custom' : type;

        switch (typeComponent) {
          case RBField:
            return RBFieldAssignProps({...props, values, errors, didSubmbit, handleChange}, false);

          case RBFieldArray:
            return RBArrayAssignProps({...props, values, errors, didSubmbit, handleChange, initialValues}, false);

          case 'custom':
            return RBCustomAssignProps({...props, values, errors, didSubmbit, handleChange}, false);

          default:
            return;
        }
      })}
      {getSubmit(customSubmit, {isLoadingSubmit, errorSubmit, successSubmit, didSubmbit, isValid, validationErrorText})}
    </Form>
  );
};
