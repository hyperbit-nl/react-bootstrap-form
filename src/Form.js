import React, { useEffect, useReducer } from 'react';
import { Formik } from 'formik';
import { RBFormObjectRecursiveForEach } from './Setup';
import { RBBuilder } from './Builder';
import { getShapeObject } from './Validation';
import { isArray, isObject, usePrevious } from './Utils';

const filterEditable = (target, source) => {
  if (isObject(target) && isObject(source)) {
    target = {...target};

    for (const key in source) {
      if (source[key] === true) { // True means that key must be hidden, otherwise this is undefined.
        delete target[key];

      } else if (isObject(target[key])) {
        target[key] = filterEditable(target[key], source[key]);

      } else if (isArray(target[key]) && Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = target[key].map((value) => (
          filterEditable(value, source[key])
        ));
      }
    }
  }

  return target;
};

const initialStateForm = {
  didLoad: false,
  validationSchema: null,
  valuesHidden: null,
  initialValues: null,
  values: null,
  isLoadingSubmit: false,
  successSubmit: false,
  errorSubmit: null,
  fieldErrorSubmit: null,
};

const reducerForm = (state, action) => {
  switch (action.type) {
    case 'update': {
      const {validationShape, values, didChangeInitialValues, ...rest} = action.payload;

      const stateNew = {
        didLoad: true,
        validationSchema: getShapeObject(validationShape),
        ...rest,
      };

      if (didChangeInitialValues) {
        stateNew.values = values;
      }

      return {...state, ...stateNew};
    }

    case 'startSubmit': {
      return {...state, isLoadingSubmit: true};
    }

    case 'submitSuccess': {
      const response = action.payload;

      return {
        ...state,
        isLoadingSubmit: false,
        successSubmit: response,
        errorSubmit: null,
        fieldErrorSubmit: null,
      };
    }

    case 'submitFailure': {
      const error = action.payload;
      const stateNew = {isLoadingSubmit: false, successSubmit: false};

      if (error) {
        const {server: serverError, field: fieldErrors} = action.payload;

        if (fieldErrors) {
          stateNew.fieldErrorSubmit = fieldErrors;
        }

        stateNew.errorSubmit = serverError ?? 'Error';

      } else {
        stateNew.errorSubmit = 'Error';
      }

      return {...state, ...stateNew};
    }

    default:
      break;
  }
};

export const RBForm = ({
  initialValues: initialValuesForm = false,
  validationErrorText,
  customSubmit,
  onSubmit,
  children
}) => {
  const initialValuesFormPrevious = usePrevious(initialValuesForm);

  const [state, dispatch] = useReducer(reducerForm, initialStateForm);

  useEffect(() => {
    const validationShape = {};
    const valuesHidden = {};
    const initialValues = {};
    const values = {};

    RBFormObjectRecursiveForEach(
      validationShape,
      valuesHidden,
      initialValues,
      isObject(initialValuesForm) ? initialValuesForm : {},
      values,
      isArray(children) ? children: [children]
    );

    dispatch({
      type: 'update',
      payload: {
        validationShape,
        valuesHidden,
        initialValues,
        values,
        didChangeInitialValues: initialValuesFormPrevious !== initialValuesForm,
      },
    });

  }, [initialValuesForm, initialValuesFormPrevious, children]);

  const handleSubmit = (values, {resetForm}) => {
    const valuesEditable = filterEditable(values, state.valuesHidden);

    if (onSubmit) {
      dispatch({type: 'startSubmit'});

      onSubmit(valuesEditable, (response) => {
        onSubmitSuccess(resetForm, response);

      }, (error) => {
        dispatch({type: 'submitFailure', payload: error});
      });

    } else {
      onSubmitSuccess(resetForm);
    }
  };

  const onSubmitSuccess = (resetForm, response = 'Success') => {
    resetForm();
    dispatch({type: 'submitSuccess', payload: response});
  };

  return (
    state.didLoad && (
      <Formik
        validationSchema={state.validationSchema}
        initialValues={state.values}
        onSubmit={handleSubmit}
        enableReinitialize
      >
        {(form) => (
          <RBBuilder
            {...{
              form,
              validationSchema: state.validationSchema,
              initialValues: state.initialValues,
              customSubmit,
              isLoadingSubmit: state.isLoadingSubmit,
              errorSubmit: state.errorSubmit,
              fieldErrorSubmit: state.fieldErrorSubmit,
              successSubmit: state.successSubmit,
              validationErrorText,
            }}
          >
            {children}
          </RBBuilder>
        )}
      </Formik>
    )
  );
};
