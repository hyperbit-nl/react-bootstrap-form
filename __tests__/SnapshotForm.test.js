import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Button } from 'react-bootstrap';
import { RBForm } from '../src/Form';
import { RBField as Field } from '../src/Field';
import { RBFieldArray as FieldArray } from '../src/FieldArray';

it("renders default Form", () => {
  const {container} = render(
    <RBForm />
  );

  expect(container).toMatchSnapshot();
});

it("renders default Form with string as customSubmit", () => {
  const {container} = render(
    <RBForm customSubmit="Custom Submit" />
  );

  expect(container).toMatchSnapshot();
});

it("renders default Form with JSX as customSubmit", () => {
  const {container} = render(
    <RBForm customSubmit={(isWaiting, success, error) => (<Button type="submit">Send</Button>)} />
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with one text Field", () => {
  const {container} = render(
    <RBForm>
      <Field name="text" type="text" label="Text" />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with one text Field and one initialValues", () => {
  const {container} = render(
    <RBForm initialValues={{text: 'Text'}}>
      <Field name="text" type="text" label="Text" />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});
it("renders Form with one text Field with defaultValue ", () => {
  const {container} = render(
    <RBForm>
      <Field name="text" type="text" label="Text" defaultValue="Default value" />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with one text Field with defaultValue and one initialValues", () => {
  const {container} = render(
    <RBForm initialValues={{text: 'Text'}}>
      <Field name="text" type="text" label="Text" defaultValue="Default value" />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with one text Field with value and one initialValues", () => {
  const {container} = render(
    <RBForm initialValues={{text: 'Text'}}>
      <Field name="text" type="text" label="Text" value="Value" />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with two text Fields", () => {
  const {container} = render(
    <RBForm>
      <Field name="text_1" type="text" label="Text 1" />
      <Field name="text_2" type="text" label="Text 2" />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with two text Fields and one initialValues", () => {
  const {container} = render(
    <RBForm initialValues={{text_1: 'Text 1'}}>
      <Field name="text_1" type="text" label="Text 1" />
      <Field name="text_2" type="text" label="Text 2" />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with two text Fields and two initialValues", () => {
  const {container} = render(
    <RBForm initialValues={{text_1: 'Text 1', text_2: 'Text 2'}}>
      <Field name="text_1" type="text" label="Text 1" />
      <Field name="text_2" type="text" label="Text 2" />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with one array Field with one text Field", () => {
  const {container} = render(
    <RBForm>
      <FieldArray name="array">
        <Field name="text" type="text" label="Text" />
      </FieldArray>
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with one array Field with one text Field and one initialValues", () => {
  const {container} = render(
    <RBForm initialValues={{array: [{text: 'Text'}]}}>
      <FieldArray name="array">
        <Field name="text" type="text" label="Text" />
      </FieldArray>
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Form with one number Field with error", async () => {
  const {container} = render(
    <RBForm initialValues={{number: 0}}>
      <Field name="number" type="number" label="Number" min={1} />
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const error = await screen.findByText('Error');
  const errorField = screen.findByText('number must be greater than or equal to 1');

  expect(container).toMatchSnapshot();
});

it("renders Form with one number Field with custom error", async () => {
  const {container} = render(
    <RBForm initialValues={{number: 0}} validationErrorText="Not Ok">
      <Field name="number" type="number" label="Number" min={1} />
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const error = await screen.findByText('Not Ok');
  const errorField = screen.findByText('number must be greater than or equal to 1');

  expect(container).toMatchSnapshot();
});

it("renders Form with one number Field with success", async () => {
  const {container} = render(
    <RBForm initialValues={{number: 1}}>
      <Field name="number" type="number" label="Number" min={1} />
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const success = await screen.findByText('Success');

  expect(container).toMatchSnapshot();
});

it("renders Form with default success onSubmit", async () => {
  const {container} = render(
    <RBForm onSubmit={(values, success, failure) => {
      success();
    }}>
      <Field name="number" type="number" label="Number" />
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const success = await screen.findByText('Success');

  expect(container).toMatchSnapshot();
});

it("renders Form with custom success onSubmit", async () => {
  const {container} = render(
    <RBForm onSubmit={(values, success, failure) => {
      success('OK');
    }}>
      <Field name="number" type="number" label="Number" />
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const success = await screen.findByText('OK');

  expect(container).toMatchSnapshot();
});

it("renders Form with default failure onSubmit", async () => {
  const {container} = render(
    <RBForm onSubmit={(values, success, failure) => {
      failure();
    }}>
      <Field name="number" type="number" label="Number" />
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const error = await screen.findByText('Error');

  expect(container).toMatchSnapshot();
});

it("renders Form with server failure onSubmit", async () => {
  const {container} = render(
    <RBForm onSubmit={(values, success, failure) => {
      failure({server: 'Not Ok'});
    }}>
      <Field name="number" type="number" label="Number" />
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const error = await screen.findByText('Not Ok');

  expect(container).toMatchSnapshot();
});

it("renders Form with field failure onSubmit", async () => {
  const {container} = render(
    <RBForm onSubmit={(values, success, failure) => {
      failure({field: {number: 'Field wrong'}});
    }}>
      <Field name="number" type="number" label="Number" />
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const errorServer = await screen.findByText('Error');
  const errorField = screen.findByText('Field wrong');

  expect(container).toMatchSnapshot();
});

it("renders Form with server and field failure onSubmit", async () => {
  const {container} = render(
    <RBForm onSubmit={(values, success, failure) => {
      failure({server: 'Not Ok', field: {number: 'Field wrong'}});
    }}>
      <Field name="number" type="number" label="Number" />
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const errorServer = await screen.findByText('Not Ok');
  const errorField = screen.findByText('Field wrong');

  expect(container).toMatchSnapshot();
});

it("renders Form with array failure onSubmit", async () => {
  const {container} = render(
    <RBForm onSubmit={(values, success, failure) => {
      failure({field: {data: [{}, {number: 'Field wrong'}]}});
    }}>
      <FieldArray name="data" initialSize={2}>
        <Field name="number" type="number" label="Number" />
      </FieldArray>
    </RBForm>
  );

  const button = await screen.findByText('Send');
  fireEvent.click(button);
  const errorServer = await screen.findByText('Error');
  const errorField = screen.findByText('Field wrong');

  expect(container).toMatchSnapshot();
});
