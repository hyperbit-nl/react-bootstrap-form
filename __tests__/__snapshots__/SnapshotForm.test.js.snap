// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Form with array failure onSubmit 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="data.0.number"
          id="rbf_data.0.number_label"
        >
          Number
        </label>
        <input
          class="form-control is-valid"
          id="data.0.number"
          type="number"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="data.1.number"
          id="rbf_data.1.number_label"
        >
          Number
        </label>
        <input
          class="form-control is-valid"
          id="data.1.number"
          type="number"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <button
        class="mb-3 btn btn-primary"
        type="button"
      >
        Add
      </button>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-red"
        role="status"
      >
        Error
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with custom success onSubmit 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="number"
        id="rbf_number_label"
      >
        Number
      </label>
      <input
        class="form-control"
        id="number"
        type="number"
        value=""
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-green"
        role="status"
      >
        OK
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with default failure onSubmit 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="number"
        id="rbf_number_label"
      >
        Number
      </label>
      <input
        class="form-control is-valid"
        id="number"
        type="number"
        value=""
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-red"
        role="status"
      >
        Error
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with default success onSubmit 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="number"
        id="rbf_number_label"
      >
        Number
      </label>
      <input
        class="form-control"
        id="number"
        type="number"
        value=""
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-green"
        role="status"
      >
        Success
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with field failure onSubmit 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="number"
        id="rbf_number_label"
      >
        Number
      </label>
      <input
        class="form-control is-invalid"
        id="number"
        type="number"
        value=""
      />
      <div
        class="invalid-feedback"
      >
        Field wrong
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-red"
        role="status"
      >
        Error
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with one array Field with one text Field 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="array.0.text"
          id="rbf_array.0.text_label"
        >
          Text
        </label>
        <input
          class="form-control"
          id="array.0.text"
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <button
        class="mb-3 btn btn-primary"
        type="button"
      >
        Add
      </button>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Form with one array Field with one text Field and one initialValues 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="array.0.text"
          id="rbf_array.0.text_label"
        >
          Text
        </label>
        <input
          class="form-control"
          id="array.0.text"
          type="text"
          value="Text"
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <button
        class="mb-3 btn btn-primary"
        type="button"
      >
        Add
      </button>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Form with one number Field with custom error 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="number"
        id="rbf_number_label"
      >
        Number
      </label>
      <input
        class="form-control is-invalid"
        id="number"
        min="1"
        type="number"
        value="0"
      />
      <div
        class="invalid-feedback"
      >
        number must be greater than or equal to 1
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-red"
        role="status"
      >
        Not Ok
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with one number Field with error 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="number"
        id="rbf_number_label"
      >
        Number
      </label>
      <input
        class="form-control is-invalid"
        id="number"
        min="1"
        type="number"
        value="0"
      />
      <div
        class="invalid-feedback"
      >
        number must be greater than or equal to 1
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-red"
        role="status"
      >
        Error
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with one number Field with success 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="number"
        id="rbf_number_label"
      >
        Number
      </label>
      <input
        class="form-control"
        id="number"
        min="1"
        type="number"
        value="1"
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-green"
        role="status"
      >
        Success
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with one text Field 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text"
        id="rbf_text_label"
      >
        Text
      </label>
      <input
        class="form-control"
        id="text"
        type="text"
        value=""
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Form with one text Field and one initialValues 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text"
        id="rbf_text_label"
      >
        Text
      </label>
      <input
        class="form-control"
        id="text"
        type="text"
        value="Text"
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Form with one text Field with defaultValue  1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text"
        id="rbf_text_label"
      >
        Text
      </label>
      <input
        class="form-control"
        id="text"
        type="text"
        value="Default value"
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Form with one text Field with defaultValue and one initialValues 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text"
        id="rbf_text_label"
      >
        Text
      </label>
      <input
        class="form-control"
        id="text"
        type="text"
        value="Text"
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Form with one text Field with value and one initialValues 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text"
        id="rbf_text_label"
      >
        Text
      </label>
      <input
        class="form-control"
        id="text"
        type="text"
        value="Value"
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Form with server and field failure onSubmit 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="number"
        id="rbf_number_label"
      >
        Number
      </label>
      <input
        class="form-control is-invalid"
        id="number"
        type="number"
        value=""
      />
      <div
        class="invalid-feedback"
      >
        Field wrong
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-red"
        role="status"
      >
        Not Ok
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with server failure onSubmit 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="number"
        id="rbf_number_label"
      >
        Number
      </label>
      <input
        class="form-control is-valid"
        id="number"
        type="number"
        value=""
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
      <span
        class="ml-2 rbf-text-red"
        role="status"
      >
        Not Ok
      </span>
    </div>
  </form>
</div>
`;

exports[`renders Form with two text Fields 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text_1"
        id="rbf_text_1_label"
      >
        Text 1
      </label>
      <input
        class="form-control"
        id="text_1"
        type="text"
        value=""
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text_2"
        id="rbf_text_2_label"
      >
        Text 2
      </label>
      <input
        class="form-control"
        id="text_2"
        type="text"
        value=""
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Form with two text Fields and one initialValues 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text_1"
        id="rbf_text_1_label"
      >
        Text 1
      </label>
      <input
        class="form-control"
        id="text_1"
        type="text"
        value="Text 1"
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text_2"
        id="rbf_text_2_label"
      >
        Text 2
      </label>
      <input
        class="form-control"
        id="text_2"
        type="text"
        value=""
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Form with two text Fields and two initialValues 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text_1"
        id="rbf_text_1_label"
      >
        Text 1
      </label>
      <input
        class="form-control"
        id="text_1"
        type="text"
        value="Text 1"
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="text_2"
        id="rbf_text_2_label"
      >
        Text 2
      </label>
      <input
        class="form-control"
        id="text_2"
        type="text"
        value="Text 2"
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders default Form 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders default Form with JSX as customSubmit 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <button
      class="btn btn-primary"
      type="submit"
    >
      Send
    </button>
  </form>
</div>
`;

exports[`renders default Form with string as customSubmit 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Custom Submit
      </button>
    </div>
  </form>
</div>
`;
